js/src/path-test.js

```diff
@@ -8,100 +8,277 @@ import {assert} from 'chai';
 import {suite, test} from 'mocha';
 import {equals} from './compare.js';
 
-import Path from './path.js';
-import type Value from './value.js';
+import {getHash} from './get-hash.js';
 import List from './list.js';
 import Map from './map.js';
+import {default as Path} from './path.js';
+import Ref from './ref.js';
+import Set from './set.js';
+import type Value from './value.js';
 import {newStruct} from './struct.js';
 
-suite('Path', () => {
+function strify(s) {
+  return JSON.stringify(s);
+}
 
-  async function assertPathEqual(expect: any, ref: Value, path: Path):
-      Promise<void> {
-    // $FlowIssue: need to be able to pass in null for ref
-    const actual = await path.resolve(ref);
-    if (actual === undefined || expect === undefined) {
-      assert.strictEqual(expect, actual);
-      return;
-    }
+function hashIdx(v: Value): string {
+  return `[#${getHash(v).toString()}]`;
+}
 
-    assert.isTrue(equals(expect, actual));
+async function assertResolvesTo(expect: ?Value, ref: Value, str: string) {
+  const p = Path.parse(str);
+  const actual = await p.resolve(ref);
+  if (expect == null) {
```

Comment:
The problem with this (and you will see this in the patchset) is that I need to add in a bunch of places:
```
let p = ...;
invariant(p !== undefined); // flow happy
if (p === null) {
  // actual null case
}
```
because without the `invariant` flow thinks - correctly, I suppose - that within the `=== null` block it can be `undefined`.

oh well.